### PR TITLE
feat(frontend): dapps promo banner

### PR DESCRIPTION
**Pull request by loki344**
# Motivation

Implement DappPromoBanner 

# Changes

Add new component
Implement Button sizes/padding/text-size/radius according to figma design
remove smallPadding in favor of more versatile size property

# Tests

No tests added

Preview:
Bildschirmfoto 2024-10-15 um 08 31 49

### CODE REVIEW

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:19]: - Please add the missing informaiton to avoid tracking. See for example the component `ExternalLink`.

- A11y information are missing

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:20]: As far as I know, a link should not correct a button. I think this is not conform with the web specification.

File: src/frontend/src/lib/components/ui/Button.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/ui/Button.svelte:6]: Please provide this in another PR

File: src/frontend/src/lib/components/ui/ButtonHero.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/ui/ButtonHero.svelte:9]: Why is the something related to the "Hero" pane modified in this PR?

**Review by peterpeterparker — CHANGES_REQUESTED** [review]
Thanks for the PR. I'm just blocking it for now because I might have some more comments after the first issue is addressed.

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:10]: I think you can reuse `Img` component for this (it defaults to "lazy")

File: src/frontend/src/lib/styles/global/button.scss
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/styles/global/button.scss:1]: Careful to remove properties here, because these classes are still not all concentrated in `Button` component, but they are quite sparse

File: tailwind.config.js
Comment (AntonioVentilii-DFINITY) [inline tailwind.config.js:56]: as discussed offline, let's keep it as primitive colors for now, and work on the mapping separately

**Review by AntonioVentilii-DFINITY — COMMENTED** [review]
Regrading the images/icons, should we add some reference/credit/source in them? How is that handled? See for example what we write in the icons

**Review by peterpeterparker — COMMENTED** [review]
`kong-swap.png` is 83.9 KB. Can its size or encoding be improved or the image provided is already tailored for the best?

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:5]: I was referencing the figma design component. 
image

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:14]: I'll use the `Img` component

File: src/frontend/src/data/dapps.json
Comment (loki344) [reply src/frontend/src/data/dapps.json:1]: As this list is 'curated' i do not think it should be edited by external contributors.

File: src/frontend/src/lib/components/ui/ButtonHero.svelte
Comment (loki344) [reply src/frontend/src/lib/components/ui/ButtonHero.svelte:9]: Because of the Button changes. Buttons do not match the figma at all at the moment. To match this Promocard to the figma design, the Button need adjustments to match the figma design.

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:16]: Mostly because I think it might make sense looking at the overall page. But I have no hard feelings changing it to p / div. Whatever you prefer. 

image

**Comment by loki344** [comment]
> Regrading the images/icons, should we add some reference/credit/source in them? How is that handled? See for example what we write in the icons

I can add the link to figma to the icons if that's what you mean..?

File: src/frontend/src/data/dapps.json
Comment (peterpeterparker) [reply src/frontend/src/data/dapps.json]: > As this list is 'curated' i do not think it should be edited by external contributors.

Good to know.

I've got one more question unrelated to the code about this: is the JSON data suppose to contains any kind of dApps on the IC or solely those that interact with Oisy through the signer standards?

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:7]: Gotcha! Double checked, it's `Dapp` in Internet Identity therefore it's consistent naming. Thanks.

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte]: There are better specialists in nomenclature at the foundation, like @meodai, but maybe for the banner an `h2` for `KongSwap` and some sort of tag for its `Featured`, assuming that after the title it's the most important information? But yeah not sure. Not a blocker.

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:11]: A `nonNullish` is missing here (we are using this as convention for null check)

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:13]: Isn't a screenshot a must to be showcased in the promo banner?

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:19]: nitpick: just sharing in case you like it, but if you use `@const` you can destruct the property.

```
{@const {logo: src} = dApp}

<Img {src} />
```

Again, not a call to action.

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:26]: @AntonioVentilii-DFINITY do we have already a maybe "Button link" component something that can potentially be used here?

File: src/frontend/src/lib/types/dapp.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/types/dapp.ts:2]: This will be replaced by some zod types once we implement the JSON file, correct? Maybe a TODO?

File: src/frontend/src/data/dapps.json
Comment (loki344) [reply src/frontend/src/data/dapps.json]: My understanding is, that we only include dApps that we want to have on the explore page

File: src/frontend/src/lib/components/ui/ButtonHero.svelte
Comment (loki344) [reply src/frontend/src/lib/components/ui/ButtonHero.svelte]: I'll make a separate PR

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:11]: thanks

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:13]: Yeah that's a good point, I'll add a separate property to make it more clear

File: src/frontend/src/lib/types/dapp.ts
Comment (loki344) [reply src/frontend/src/lib/types/dapp.ts:2]: Ah, no then let me add the zod schema directly

File: src/frontend/src/lib/components/ui/ButtonHero.svelte
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/components/ui/ButtonHero.svelte]: I forgot to send the message, but yes, I would prefer another PR too, and if necessary for this, to be done first. It should be a quick review I imagine

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:25]: Correct! We have `ExternalLinkAsIcon` that is a derivative of Button

Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:25]: I think we can resolve this issue, I clarified with artem, this button will not directly navigate. Instead it will open the modal. Will change it to a button

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:18]: You might want to revisit the `alt` here. The app name isn't really an accessible text I would say. It should probably be a text that contains the app name.

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:25]: This will opens the modal you mentioned right? Should we maybe add a TODO in some sort?

File: src/frontend/src/lib/types/dapp.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/types/dapp.ts:4]: Thanks. As I mentionned in my very first comment on the review, it would be nice to provide the type / zod object in a separate PR for separation of converns. I can review it here but for future PR, please try to split logic and presentation.

File: src/frontend/src/lib/types/dapp.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/types/dapp.ts:8]: can you add a TODO to validate that it's HTTPs

File: src/frontend/src/lib/types/dapp.ts
Comment (peterpeterparker) [inline src/frontend/src/lib/types/dapp.ts:7]: There seems to be some sort of min and max logic implemented in the website. We should try to replicate those rules. Does not need to be in this PR so maybe you can add a TODO referencing the code of the portal: https://github.com/dfinity/portal/blob/34a0328ed4792f5a7f3943be73f13f5abaefb4b8/plugins/validate-showcase.js#L179

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte]: I honestly did not understood that the property you added made it clear. Also not sure that's the appropriate solution.

For this PR maybe let's make this obvious on the type of the property and let's iterate in another PR.

I did not got it. Should rather be the type of `Dapp` be improved to make this requirement obvious?

```
export let dapp: Omit<Dapp, "background"> & Required<Pick<Dapp, "background">>
```

WDYT?

**Review by peterpeterparker — COMMENTED** [review]
Probably the last round of comments, thanks in advance for the feedbacks.

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:25]: I can, although I'm implementing this right now xD

Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:25]: On second thought, no I don't think this is necessary. This component does actually not care what the caller binds to onclick

File: src/frontend/src/lib/types/dapp.ts
Comment (AntonioVentilii-DFINITY) [inline src/frontend/src/lib/types/dapp.ts:34]: @peterpeterparker @loki344 what do you think of renaming this `DAppInfo` or something like that?

Comment (peterpeterparker) [reply src/frontend/src/lib/types/dapp.ts:34]: In Internet Identity they actually use `DappDescription` for the JSON data and a wrapper `KnownDapp` for those that have been asserted given that they fetch the list from the portal. 

So maybe `DappDescription` as well?

Ultimately we can use the same zod parser in II and Oisy and Portal if it's the same structure and if the foundation wants to avoid another mess.

II source code: https://github.com/dfinity/internet-identity/blob/a537c50f9707034ff4d2cfc41d77ff8caa0a845e/src/frontend/src/flows/dappsExplorer/dapps.ts#L15C13-L15C28

Comment (loki344) [reply src/frontend/src/lib/types/dapp.ts:34]: IMHO adding things like "Info" or "Data" to variable names rarely adds any value. I will use DappDescription, even though I think the interface is more than just a description (the field description inside of it implies that the description is part of the interface).

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:13]: We gonna have to review this because even if we made screenshots mandatory in the property, it can still be empty. If you want to improve the types before merge it's nice but, we can do it afterwards too to move on.

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [inline src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:14]: Just a note for myself: we probably also gonna need to review this and we might need an `object-fit: cover;` but let's move on.

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (peterpeterparker) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte:34]: Can you comment the all button for now before merge please and add a TODO?

**Review by peterpeterparker — APPROVED** [review]
LGTM.

Please comment the unused button before merge and wait for @AntonioVentilii-DFINITY approval given that he also participated to the review. Thanks!

File: src/frontend/src/lib/components/dapps/DappPromoBanner.svelte
Comment (loki344) [reply src/frontend/src/lib/components/dapps/DappPromoBanner.svelte]: done

**Review by AntonioVentilii-DFINITY — APPROVED** [review]

**Pull request merged**
